src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -248,7 +248,10 @@ private static void ReturnBlocks(MemoryPoolBlock2 block)
 
         private void ScheduleWrite()
         {
-            _thread.Post(_this => _this.WriteAllPending(), this);
```

Comment:
I don't see the need for `_writeQueueState` and all this Interlocked stuff.

We could simply create a `_writeQueued` bool instead and set it and `scheduleWrite` to true inside of `WriteAsync` with the `_contextLock` taken if `!_writeQueued && immediate`. Then you simply set `_writeQueued` to false inside of `WriteAllPending` where you once again have the `_contextLock`.